sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UpdateNullability.scala

```diff
@@ -52,3 +52,22 @@ object UpdateAttributeNullability extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * Updates nullability of [[GroupingExprRef]]s in a resolved LogicalPlan by using the nullability of
+ * referenced grouping expression.
+ */
+object UpdateGroupingExprRefNullability extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
```

Comment:
This is an optimizer rule, shall we use `transformUp`?

Reply:
Sure, fixed in https://github.com/apache/spark/pull/31913/commits/2e79eb910f5d4ef60227f7352e86dbed2786cb7a